### PR TITLE
Profile import identifier handling

### PR DESCRIPTION
**Pull request by david-waltermire-nist**
# Committer Notes

Provides identifier reassignment capabilities for remapping identifiers in imported catalogs during the import phase of profile resolution.

### All Submissions:

- [ ] Have you selected the correct base branch per [Contributing](https://github.com/usnistgov/OSCAL/blob/main/CONTRIBUTING.md) guidance?
- [ ] Have you set "[Allow edits and access to secrets by maintainers
](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork)"?
- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/usnistgov/OSCAL/pulls) for the same update/change?
- [ ] Have you squashed any non-relevant commits and commit messages? \[[instructions](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History)\]
- [ ] Do all automated CI/CD checks pass?

By submitting a pull request, you are agreeing to provide this contribution under the [CC0 1.0 Universal public domain](https://creativecommons.org/publicdomain/zero/1.0/) dedication.

### Changes to Core Features:

- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your core changes, as applicable?
- [ ] Have you included examples of how to use your new feature(s)?
- [ ] Have you updated all [OSCAL website](https://pages.nist.gov/OSCAL) and readme documentation affected by the changes you made? Changes to the OSCAL website can be made in the docs/content directory of your branch.

### CODE REVIEW

**Comment by david-waltermire-nist** [comment]
The following is an example of the resulting solution.

```json
{
    "imports": [
        {
            "href": "#FedRAMP_rev4_MODERATE-baseline_profile.json",
            "include-controls": [
            {
                "with-ids": [
                "ac-6.9",
                "ac-6.10",
                "ac-7",
                "ac-8"
                ]
            }
            ],
            "identifier-reassignment": {
                "with-prefix": "prefixA-",
                "with-suffix": "-suffixA",
                "controls": {
                    "with-prefix": "prefixB-",
                    "with-suffix": "-suffixB",
                    "reassignments": [
                        {
                            "from": "ac-6.9",
                            "to": "ac-6.9-new"
                        }
                    ]
                }
                // ac-6.9 -> prefixA-prefixB-ac-6.9-new-suffixB-suffixA
                // ac-6.10 -> prefixA-prefixB-ac-6.10-suffixB-suffixA
                // ac-7 -> prefixA-prefixB-ac-7-suffixB-suffixA
                // ac-8 -> prefixA-prefixB-ac-8-new-suffixB-suffixA
            }
        }
    ]
}
```

**Comment by AdamOlineCB** [comment]
When reading through the example, I see "with-ids" acting as a filter of sorts in the "include-controls" section, identifying the controls to include from the profile, and then I scan down to the "identifier-reassignment" that uses "with-prefix" and "with-suffix" and my first thought was "are those also filters?". After studying for a few more minutes and reviewing the examples it is obvious they are actions to apply the prefix and suffix values for the resulting identifiers.

My suggestion would be to change the "with-" portion to something that more clearly communicates these are actions, such as "use-" or "apply-", e.g.:

"use-prefix": "prefixB-",
"use-suffix": "-suffixB",

or

**Comment by aj-stein-nist** [comment]
> ```
> "apply-prefix": "prefixB-",
> "apply-suffix": "-suffixB",
> ```
> 
> I think I like apply- best of these two.

I too like the `apply-` and the rationale Adam provides resonated with. I think it is looking at too much Python using `with` context managers. 😆

**Comment by aj-stein-nist** [comment]
> ```json
> {
>     "imports": [
>         {
>             "href": "#FedRAMP_rev4_MODERATE-baseline_profile.json",
>             "include-controls": [
>             {
>                 "with-ids": [
>                 "ac-6.9",
>                 "ac-6.10",
>                 "ac-7",
>                 "ac-8"
>                 ]
>             }
>             ],
>             "identifier-reassignment": {
>                 "with-prefix": "prefixA-",
>                 "with-suffix": "-suffixA",
>                 "controls": {
>                     "with-prefix": "prefixB-",
>                     "with-suffix": "-suffixB",
>                     "reassignments": [
>                         {
>                             "from": "ac-6.9",
>                             "to": "ac-6.9-new"
>                         }
>                     ]
>                 }
>                 // ac-6.9 -> prefixA-prefixB-ac-6.9-new-suffixB-suffixA
>                 // ac-6.10 -> prefixA-prefixB-ac-6.10-suffixB-suffixA
>                 // ac-7 -> prefixA-prefixB-ac-7-suffixB-suffixA
>                 // ac-8 -> prefixA-prefixB-ac-8-suffixB-suffixA
>             }
>         }
>     ]
> }
> ```

I am confused by the use of `with-prefix` in a few places and the model didn't really help me. Does the nested `with-prefix` and `with-suffix` declarations apply to some controls and not others as a subset of the `include-controls` assembly? If not, why do we double up the directives? I think I am just confused by the example given all controls in the example have wrapped suffixes and prefixes the same way.

**Comment by iMichaela** [comment]
Is the following interpretation accurate, considering the following assumptions:
1) I am creating a new profile derived from FedRAMP's moderate profile, and I want to have an easy way to remember what I did since I am tailoring the controls in the process
2) I want to know (easily remember that my controls started with _fedramp_ baseline, _moderate_ impact, and will end in my profile for highly regulated env _hre_, _low_ impact profile
3) I want later on to easily remember what I did in 2) just by looking at the ID

NOTE: I am using, as an example, the prefix for governance traceability and the suffix for impact level.

```
{
    "imports": [
        {
            "href": "#FedRAMP_rev4_MODERATE-baseline_profile.json",
            "include-controls": [
            {
                "with-ids": [
                "ac-6.9",
                "ac-6.10",
                "ac-7",
                "ac-8"
                ]
            }
            ],
            "identifier-reassignment": {
                "with-prefix": "mp-",
                "with-suffix": "-low",
                "controls": {
                    "with-prefix": "fedramp-",
                    "with-suffix": "-moderate",
                    "reassignments": [
                        {
                            "from": "ac-6.9",
                            "to": "hre-fedramp-ac-6.9-moderate-low" 
                        }
                    ]
                }
                // ac-6.9 -> hre-fedramp-ac-6.9-moderate-low
            }
        }
    ]
}
```
OR
The `identifier-reassignment` is only informative and the `reassignment` should read 

```
"reassignments": [
                        {
                            "from": "ac-6.9",
                            "to": "ac-6.9-new" 
                        }
                    ]
```
and the new ID is actually `ac-6.9-new` 

I am not sure I understand the part of the example in red that highlights the **new** name PLUS the chained prefixes-suffixes. Where or when should it be used as shown?

File: src/metaschema/oscal_profile_metaschema.xml [hunk 1]
```diff
@@ -76,6 +76,50 @@
                               <p>Identifies which controls to exclude, or eliminate, from the set of included controls by control identifier or match pattern.</p>
                         </remarks>
                   </assembly>
+                  <define-assembly name="identifier-reassignment">
+                        <formal-name>Mapping</formal-name>
+                        <description>Maps objects to new ID values</description>
+                        <model>
+                              <field ref="with-prefix"/>
+                              <field ref="with-suffix"/>
```
Comment (david-waltermire-nist) [inline src/metaschema/oscal_profile_metaschema.xml:84]:
```suggestion
                              <field ref="apply-prefix"/>
                              <field ref="apply-suffix"/>
```

File: src/metaschema/oscal_profile_metaschema.xml [hunk 2]
```diff
@@ -87,6 +131,53 @@
                   </import>
             </example>
       </define-assembly>
+      
+      <!-- CHANGED: Added Prefix Suffix Support for ID Reassignment -->
+
+      <define-field name="with-prefix" as-type="token" scope="local">
```
Comment (david-waltermire-nist) [inline src/metaschema/oscal_profile_metaschema.xml:137]:
```suggestion
      <define-field name="apply-prefix" as-type="token" scope="local">
```

File: src/metaschema/oscal_profile_metaschema.xml [hunk 3]
```diff
@@ -76,6 +76,50 @@
...
+                        </model>
+                        <remarks>
+                              <p>Contains an object-type set of identifier reassignment rules to be applied to related objects as an import is processed.</p>
+                              <p>During reassignment, all imported role, control, group, parameter, and part identifiers are handled as follows:</p>
+                              <ol>
+                                    <li>Identifiers are first reassigned using the <code>role</code>, <code>control</code>, <code>group</code>, <code>parameter</code>, and <code>part</code> directives matching the identifier's object type. This will determine the effective value of a given identifier. If a directive is not provided for a given object type, then the effective value is the original identifier value.</li>
+                                    <li>If a <code>with-prefix</code> is defined, the value will be added to the beginning of the effective identifier value.</li>
+                                    <li>If a <code>with-suffix</code> is defined, the value will be added to the end of the effective identifier value.</li>
```
Comment (david-waltermire-nist) [inline src/metaschema/oscal_profile_metaschema.xml:117]:
```suggestion
                                    <li>If a <code>apply-prefix</code> is defined, the value will be added to the beginning of the effective identifier value.</li>
                                    <li>If a <code>apply-suffix</code> is defined, the value will be added to the end of the effective identifier value.</li>
```

File: src/metaschema/oscal_profile_metaschema.xml [hunk 4]
```diff
@@ -87,6 +131,53 @@
...
       </define-assembly>
+      
+      <!-- CHANGED: Added Prefix Suffix Support for ID Reassignment -->
+
+      <define-field name="with-prefix" as-type="token" scope="local">
+            <formal-name>Identifier Reassignment Prefix</formal-name>
+            <description>Adds the defined value to the beginning of the reassigned identifier.</description>
+      </define-field>
+
+      <define-field name="with-suffix" as-type="token" scope="local">
```
Comment (david-waltermire-nist) [inline src/metaschema/oscal_profile_metaschema.xml:142]:
```suggestion
      <define-field name="apply-suffix" as-type="token" scope="local">
```

File: src/metaschema/oscal_profile_metaschema.xml [hunk 5]
```diff
@@ -87,6 +131,53 @@
...
+            <formal-name>Identifier Reassignment Suffix</formal-name>
+            <description>Concatenates the defined value on the end of the reassigned identifier.</description>
+      </define-field>
+
+      <define-assembly name="id-reassignment-context" scope="local">
+            <formal-name>Identifier Reassignment Context</formal-name>
+            <description>A set of rules for reassigning identifier for a specific type of object.</description>
+            <model>
+                  <field ref="with-prefix"/>
+                  <field ref="with-suffix"/>
```
Comment (david-waltermire-nist) [inline src/metaschema/oscal_profile_metaschema.xml:152]:
```suggestion
                  <field ref="apply-prefix"/>
                  <field ref="apply-suffix"/>
```

File: src/metaschema/oscal_profile_metaschema.xml [hunk 6]
```diff
@@ -87,6 +131,53 @@
...
+                              <description>A new valid identifier value for an object that is to be remapped.</description>
+                        </define-flag>
+                  </define-assembly>
+            </model>
+            <remarks>
+                  <p>During reassignment, for each identifier matching a given object type:</p>
+                  <ol>
+                        <li>Identifiers are first reassigned using the <code>from</code> and <code>to</code> tokens. An identifier that matches the <code>from</code> will be reassigned to the <code>to</code> value to establish the identifier's effective value. If a <code>from</code> does not match, the identifier's effective value is the original identifier value.</li>
+                        <li>If a <code>with-prefix</code> is defined, the value will be added to the beginning of the effective identifier.</li>
+                        <li>If a <code>with-suffix</code> is defined, the value will be added to the end of the effective identifier.</li>
```
Comment (david-waltermire-nist) [inline src/metaschema/oscal_profile_metaschema.xml:172]:
```suggestion
                        <li>If a <code>apply-prefix</code> is defined, the value will be added to the beginning of the effective identifier.</li>
                        <li>If a <code>apply-suffix</code> is defined, the value will be added to the end of the effective identifier.</li>
```

File: src/specifications/profile-resolution/profile-resolution-specml.xml [hunk 7]
```diff
@@ -419,17 +419,25 @@ intermediate:
           <src>mapping</src> child of a given
           <src>import</src> provides a simple ID remapping for objects included from that specific import. This provides the means for profile authors to proactively avoid clashing IDs of controls and other objects.
         </p>
-        <p>The Mapping section consists of 5 optional subsections, each covering a particular type of object. Each subsection is a list of ID mappings to be applied for objects that are the parent object type.</p>
+        <p>The Mapping section consists of 3 optional subsections, each covering a particular type of object. The valid subsections are as follows: </p>
+        <ul>
+          <li><p>Control</p></li>
+          <li><p>Group</p></li>
+          <li><p>Resource</p></li>
```
Comment (david-waltermire-nist) [inline src/specifications/profile-resolution/profile-resolution-specml.xml:426]:
```suggestion
          <li><p>role</p></li>
          <li><p>control</p></li>
          <li><p>group</p></li>
          <li><p>parameter</p></li>
          <li><p>part</p></li>
```

**Comment by david-waltermire-nist** [comment]
This work will be completed once #1474 is completed.

**Comment by aj-stein-nist** [comment]
@david-waltermire-nist I would like to take some time next week to meet up about this work and how it relates to other issues. After that, I will determine how we work on the issue in future sprints and update this branch.

**Pull request closed without merging**